crates/verona-rt/src/cown.rs

```diff
@@ -11,50 +11,115 @@ use verona_rt_sys as ffi;
 /// You can't return them by value over an FFI boundry, as the ABI will be wrong, and you'll
 /// end up
 
-pub struct CownPtr(pub(crate) ffi::CownPtr);
+// +-------+-----------+----------------------+
+// | Cown  | DtorThunk | Rust Manged Whatever |
+// +-------+-----------+----------------------+
+// | ActualCown        |
+// +-------------------+----------------------+
+// | CownData                                 |
+// +------------------------------------------+
 
-impl std::ops::Drop for CownPtr {
+pub struct CownPtr<T> {
+    pub(crate) ptr: ffi::CownPtr,
+    // TODO: Is this right wrt send/sync.
+    _marker: PhantomData<CownData<T>>,
+}
+
+impl<T> CownPtr<T> {
+    fn cown_data(&self) -> *mut CownData<T> {
+        self.ptr.addr() as _
+    }
+
+    /// Safety: lol
+    #[cfg(test)]
+    unsafe fn yolo_data(&mut self) -> &mut T {
+        &mut (*self.cown_data()).data
+    }
+}
+
+#[repr(C)]
+struct ActualCown {
+    _marker: std::mem::MaybeUninit<[u64; 4]>,
+}
+
+#[repr(C)]
+pub(crate) struct CownData<T> {
+    // Must be first, so we can convert pointers between the two.
+    cown: ActualCown,
+    pub data: T,
+}
+
+impl<T> std::ops::Drop for CownPtr<T> {
     fn drop(&mut self) {
-        unsafe { ffi::cown_int_delete(&mut self.0) };
+        unsafe { ffi::boxcar_cownptr_drop(&mut self.ptr) };
     }
 }
 
-impl Clone for crate::cown::CownPtr {
-    fn clone(&self) -> crate::cown::CownPtr {
+impl<T> Clone for crate::cown::CownPtr<T> {
+    fn clone(&self) -> Self {
         unsafe {
             let mut new = mem::zeroed();
-            ffi::cown_int_clone(&self.0, &mut new);
-            Self(new.assume_init())
+            ffi::boxcar_cownptr_clone(&self.ptr, &mut new);
+            Self {
+                ptr: new,
+                _marker: PhantomData,
+            }
         }
     }
 }
 
-impl CownPtr {
+extern "C" fn drop_glue<T>(cown: *mut ()) {
+    let cown = cown as *mut CownData<T>;
+
+    unsafe {
+        ptr::drop_in_place(ptr::addr_of_mut!((*cown).data));
+    }
+}
+
+impl<T> CownPtr<T> {
     /// Must be inside a runtime.
     // TODO: Enforce that.
-    pub fn new(value: i32) -> Self {
+    pub fn new(value: T) -> Self {
         unsafe {
             // The C++ code called here will read from the old value of cown to attempt to free it.
             // Luckely for us, `nullptr` is a valid value for a cown_ptr, and we can create one easily.
-            let mut cown = mem::zeroed();
-            ffi::cown_int_new(value, &mut cown);
-            Self(cown.assume_init())
+            let mut cown_ptr = mem::zeroed();
+
+            ffi::boxcar_cownptr_new(
+                // TODO: Why is this needed? AAAH. +8 doesn't work.
```

Comment:
With `+9` removed:
```
$ cargo run --example cown-minimal
   Compiling verona-rt v0.1.0 (/home/alona/icnotes/60010-final-year-project/crates/verona-rt)
    Finished dev [unoptimized + debuginfo] target(s) in 0.64s
     Running `/home/alona/icnotes/60010-final-year-project/target/debug/examples/cown-minimal`
[crates/verona-rt/src/cown.rs:100] &this.ptr = CownPtr(
    0x00007f2c9ed40010,
)
Segmentation fault
```
as-is:
```
$ cargo run --example cown-minimal
   Compiling verona-rt v0.1.0 (/home/alona/icnotes/60010-final-year-project/crates/verona-rt)
    Finished dev [unoptimized + debuginfo] target(s) in 0.63s
     Running `/home/alona/icnotes/60010-final-year-project/target/debug/examples/cown-minimal`
[crates/verona-rt/src/cown.rs:100] &this.ptr = CownPtr(
    0x00007f295c660010,
)
$
```
So it doesn't look like alighment

Reply:
Intregue: replacing the `+9` with adding `[u8;9] to the end of `CownData` means we still fail: Potentially we're doing something dodgy with dereferncing the entire struct.

This would also explain why it works with `--release`

Reply:
False lead: turns out it was a 16byte object header that came before the allocated pointer (but needed to be factored into the size of the allocation). Adding 9 worked because snmalloc rouned up the size to be a multiple of 8 (on x64).

Solved in https://github.com/aDotInTheVoid/boxcars/commit/c2376eb46a3cef08a8d83d9e856a2709375c8d29.

https://github.com/aDotInTheVoid/boxcars/blob/c2376eb46a3cef08a8d83d9e856a2709375c8d29/docs/layout.md#actualcown-shenanigans has info on what the layout looks like.